File: Day157_2025/Q157_Leetcode.py
class Solution:
    def robotWithString(self, s: str) -> str:
        from collections import Counter

        count = Counter(s)
        stack = []
        res = []
        min_char = 'a'

        for c in s:
            stack.append(c)
            count[c] -= 1

            while min_char <= 'z' and count[min_char] == 0:
                min_char = chr(ord(min_char) + 1)

            while stack and stack[-1] <= min_char:
                res.append(stack.pop())

        return ''.join(res)

File: Day157_2025/Q157_Leetcode.java
import java.util.*;

class Q157_Leetcode {
    public String robotWithString(String s) {
        int[] count = new int[26];
        for (char c : s.toCharArray()) {
            count[c - 'a']++;
        }

        Stack<Character> stack = new Stack<>();
        StringBuilder res = new StringBuilder();
        char minChar = 'a';

        for (char c : s.toCharArray()) {
            stack.push(c);
            count[c - 'a']--;

            while (minChar <= 'z' && count[minChar - 'a'] == 0) {
                minChar++;
            }

            while (!stack.isEmpty() && stack.peek() <= minChar) {
                res.append(stack.pop());
            }
        }

        return res.toString();
    }
}
